src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/payload/history/HaFlowDumpPayload.java

```diff
@@ -63,6 +65,10 @@ public class HaFlowDumpPayload implements Serializable {
     List<HaSubFlowPayload> haSubFlows;
     HaFlowPathPayload forwardPath;
     HaFlowPathPayload reversePath;
+    @JsonInclude(value = JsonInclude.Include.CUSTOM,
+            valueFilter = JsonIncludeObjectHavingNonNullFieldOrNonEmptyList.class)
     HaFlowPathPayload protectedForwardPath;
+    @JsonInclude(value = JsonInclude.Include.CUSTOM,
+            valueFilter = JsonIncludeObjectHavingNonNullFieldOrNonEmptyList.class)
```

Comment:
Just for the information to consider:
Have you thought about to move the logic of setting the values for the protectedForwardPath and protectedReversePath fields to the HaFlowHistoryMapper? I mean if we have not null, but completely empty inside objects, do we actually need to instantiate them at all? what if prevent the instantiation of such objects?
I think the best place to to it is the HaFlowHistoryMapper. or HistoryService.getHaFlowHistoryEvents() method.
As result we do need to implement JsonIncludeObjectHavingNonNullFieldOrNonEmptyList with the reflection hacks of setting the private fields accessible true.
:)

Reply:
Just another argument to have work-in-progress code reviews, because I also thought and tried it first via the mapper. It doesn't work unfortunately. Even if we don't create object at all, Jackson still renders the model according to what we have in this POJO by default including all null fields. Basically, POJO defines a model with its metadata, doesn't matter what's inside.
There is a mechanism to not display null or empty fields using `JsonInclude`, but it is tricky. The collections are never null for Jackson, so if you have a nested object with a list, this object always has a member: a collection, even if it is empty. If you put the appropriate annotation on that field, it hides that field, but still consider the parent object as non-empty. So, the object is going to be displayed, but with a hidden collection. There is no built-in annotation `NON_EMPTY` but for objects, so I had to implement it.